### PR TITLE
Error reporting via Sentry

### LINKED ISSUE #37: Error Reporting via Sentry
We have a sentry backend avaialble, credentials can be provided on request

### LINKED ISSUE #105: Redesign the settings screen
The settings screen highlights some options that are very unlikely to be used (e.g. licenses and privacy policy). We should redesign this screen and e.g. remove their icons and decrease font size. A great example of a beautiful settings screen from the HPI-App:
image

### LINKED ISSUE #199: Redesign settings screen

### PR DESCRIPTION
**Pull request by JonasWanke**
**Closes: #37, closes: #199, closes: #105**

We now report errors to Sentry. Other events can also be reported using `reportEvent`.

This also adds the current `schoolId` as a query parameter to API requests for better server performance.

TODO:
- [x] More detailed testing when I have an account ([SC-3747](https://ticketsystem.schul-cloud.org/browse/SC-3747))
- [x] Test reporting of flutter errors in production (flutter/flutter#48972)

image

### CODE REVIEW

File: lib/app/error_reporting.dart [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    storage = services.storage;
+    // ignore: avoid_catching_errors
+  } on NoSuchMethodError {
+    // GetIt always asserts that a factory is already registered for that type.
+    // In production code, there is no specific exception being thrown (and of
+    // course no assertions), but it will crash by calling methods on the
+    // non-existent (`null`) factory.
+  }
+  // We don't have the permission to report errors or storage is null and we
+  // have a Schrödinger's cat-situation in which we can't guarantee a permission
```
Comment (MarcelGarus) [inline lib/app/error_reporting.dart]:
```suggestion
  // have a Schrödinger's cat situation in which we can't guarantee a permission
```

File: lib/app/services/api_network.dart [hunk 2]
```diff
@@ -58,10 +51,25 @@ class ApiNetworkService {
...
-  /// Makes an http delete request to the api.
+  /// Makes an HTTP DELETE request to the api.
   Future<http.Response> delete(String path) {
     return _network.delete(_url(path), headers: _getHeaders());
   }
+
+  String _url(String path) {
+    assert(path != null);
+    assert(path.isNotEmpty);
+    assert(path[0] != '/');
```
Comment (MarcelGarus) [inline lib/app/services/api_network.dart]:
```suggestion
    assert(!path.startsWith('/'));
```

File: lib/main.dart [hunk 3]
```diff
@@ -61,57 +60,64 @@ const schulCloudAppConfig = AppConfig(
...
-          return Container(
-            color: Colors.white,
-            alignment: Alignment.center,
-            child: CircularProgressIndicator(),
-          );
+    logger.d('Registering remaining services…');
+    services
+      ..registerSingletonAsync((_) async {
+        // We need to initialize TimeMachine before launching the app, and using
+        // get_it to keep track of initialization statuses is the simplest way.
```
Comment (MarcelGarus) [inline lib/main.dart]:
```suggestion
        // GetIt to keep track of initialization statuses is the simplest way.
```
same as in `error_reporting.dart`

File: lib/settings/widgets/settings_screen.dart [hunk 4]
```diff
@@ -3,55 +3,92 @@ import 'package:schulcloud/app/app.dart';
...
+          trailing: Icon(Icons.open_in_new),
+        ),
+        ListTile(
+          onTap: () => tryLaunchingUrl('mailto:info@schul-cloud.org'),
+          leading: Icon(Icons.mail_outline),
+          title: Text(s.settings_about_contact),
+          trailing: Icon(Icons.open_in_new),
+        ),
+        LegalBar(),
+      ]),
```
Comment (MarcelGarus) [inline lib/settings/widgets/settings_screen.dart]:
```suggestion
      ],),
```
\+ dartfmt

**Review by MarcelGarus — CHANGES_REQUESTED** [review]

**Review by MarcelGarus — APPROVED** [review]

**Comment by JonasWanke** [comment]
I was able to test the reported errors with Sentry; exceptions seem to all get caught. Render overflows are not yet reported, maybe we should open an issue for that? And we might want to provide a custom [`ErrorWidget`](https://api.flutter.dev/flutter/widgets/ErrorWidget-class.html) to flutter that better fits our app.

**Review by MarcelGarus — APPROVED** [review]

**Pull request merged**